internal/worker/purger.go

```diff
@@ -30,21 +34,60 @@ func NewPurger(poolSize int, acrClient api.AcrCLIClientInterface, loginURL strin
 }
 
 // PurgeTags purges a list of tags concurrently, and returns a count of deleted tags and the first error occurred.
-func (p *Purger) PurgeTags(ctx context.Context, tags *[]acr.TagAttributesBase) (int, error) {
-	jobs := make([]job, len(*tags))
-	for i, tag := range *tags {
-		jobs[i] = newPurgeTagJob(p.loginURL, p.repoName, p.acrClient, *tag.Name)
-	}
+func (p *Purger) PurgeTags(ctx context.Context, tags []acr.TagAttributesBase) (int, error) {
+	var deletedTags atomic.Int64 // Count of successfully deleted tags
+	group := p.pool.NewGroup()
+	for _, tag := range tags {
+		group.SubmitErr(func() error {
+			resp, err := p.acrClient.DeleteAcrTag(ctx, p.repoName, *tag.Name)
+			if err == nil {
+				fmt.Printf("Deleted %s/%s:%s\n", p.loginURL, p.repoName, *tag.Name)
+				// Increment the count of successfully deleted tags atomically
+				deletedTags.Add(1)
+				return nil
+			}
+
+			if resp != nil && resp.Response != nil && resp.StatusCode == http.StatusNotFound {
+				// If the tag is not found it can be assumed to have been deleted.
+				deletedTags.Add(1)
+				fmt.Printf("Skipped %s/%s:%s, HTTP status: %d\n", p.loginURL, p.repoName, *tag.Name, resp.StatusCode)
+				return nil
+			}
 
-	return p.process(ctx, &jobs)
+			fmt.Printf("Failed to delete %s/%s:%s, error: %v\n", p.loginURL, p.repoName, *tag.Name, err)
+			return err
```

Comment:
instead of exiting, can we just keep track of all tags that couldn't be deleted w/ their error and then just return that?

Reply:
https://github.com/Azure/acr-cli/issues/471